src/platforms/esp32/components/avm_builtins/i2c_driver.c

```diff
@@ -613,6 +638,54 @@ static NativeHandlerResult i2cdriver_consume_mailbox(Context *ctx)
     return cmd == I2CCloseCmd ? NativeTerminate : NativeContinue;
 }
 
+// The native driver API defined here is supported only for SMP builds
+#ifndef AVM_NO_SMP
+
+static Context *get_i2c_context(term i2c_port, GlobalContext *global)
+{
+    if (UNLIKELY(!term_is_pid(i2c_port))) {
+        ESP_LOGW(TAG, "Given term is not a I2C port driver.");
+        return NULL;
+    }
+
+    int local_process_id = term_to_local_process_id(i2c_port);
+    Context *ctx = globalcontext_get_process_lock(global, local_process_id);
+
+    if (ctx->native_handler != i2cdriver_consume_mailbox || (ctx->platform_data == NULL)) {
+        ESP_LOGW(TAG, "Given term is not a I2C port driver.");
+        globalcontext_get_process_unlock(global, ctx);
+        return NULL;
+    }
+
+    return ctx;
+}
+
+I2CAcquireResult i2c_driver_acquire(term i2c_port, i2c_port_t *i2c_num, GlobalContext *global)
+{
+    Context *ctx = get_i2c_context(i2c_port, global);
+    if (IS_NULL_PTR(ctx)) {
+        return I2CAcquireInvalidPeripheral;
+    }
+    struct I2CData *i2c_data = ctx->platform_data;
```

Comment:
Process table lock is probably not
held when i2c_driver_close is called, so this can be NULL.

It is not an atomic either, so theoretically it can be garbage.

Reply:
`get_i2c_context` locks it, I can rename it for clarity.

Reply:
Does this table help?

| i2c_driver_close task                              | i2c_driver_acquire task                          |
|----------------------------------------------------|--------------------------------------------------|
| static void i2c_driver_close(Context *ctx)         | I2CAcquireResult i2c_driver_acquire(...) |
| {                                                  | { |
|    struct I2CData *i2c_data = ctx->platform_data;  |   Context *ctx = get_i2c_context_lock(i2c_port, global); |
|                                                    |    if (UNLIKELY(!term_is_pid(i2c_port))) { |
|                                                    |      // skipped |
|                                                    |    } |
|                                                    |    int local_process_id = term_to_local_process_id(i2c_port); |
|                                                    |    Context *ctx = globalcontext_get_process_lock(global, local_process_id); |
|                                                    |        if (IS_NULL_PTR(ctx)) { |
|                                                    |      // skipped |
|                                                    |    } |
|                                                    |    struct I2CData *i2c_data = ctx->platform_data; |
|                                                    |   return ctx; |
|    MUTEX_LOCK(i2c_data->mutex);                    |  |
|    ctx->platform_data = NULL;                      |  |
|                                                    | struct I2CData *i2c_data = ctx->platform_data; |
|                                                    | /* i2c_data is NULL */ |
|    MUTEX_UNLOCK(i2c_data->mutex);                  | |
|                                                    | MUTEX_LOCK(i2c_data->mutex); |
|                                                    | *i2c_num = i2c_data->i2c_num;  <-- crash |


or even

| i2c_driver_close task                              | i2c_driver_acquire task                          |
|----------------------------------------------------|--------------------------------------------------|
|    MUTEX_LOCK(i2c_data->mutex);                    | |
|    ctx->platform_data = NULL;                      | struct I2CData *i2c_data = ctx->platform_data; |
|                                                    | /* i2c_data is not NULL nor valid */ |

Reply:
thanks for the clarification, it pointed out the problem.
I did a quick and simple fix that should fix it, I'll check it again later but sounds reasonable.